# PRAKTIKUM 9

## Python Learning

In [1]:
print("Hello, World!")

Hello, World!


### Analisis
Kode `print("Hello, World!")` digunakan untuk menampilkan teks ke layar. Fungsi `print()` mencetak string yang diberikan, sehingga hasilnya muncul sebagai **Hello, World!** pada output.


## Python Indentation

In [2]:
if 5 > 2:
  print("Five is greater than two!")

Five is greater than two!


### Analisis
Kode tersebut adalah contoh pernyataan kondisi (if) dalam Python. Baris `if 5 > 2:` mengecek apakah angka 5 lebih besar dari 2. Karena kondisinya benar, maka perintah `print("Five is greater than two!")` dijalankan dan menampilkan teks tersebut pada output.


In [6]:
if 5 > 2:
 print("Five is greater than two!") 
if 5 > 2:
        print("Five is greater than two!") 

Five is greater than two!
Five is greater than two!


### Analisis
Kode tersebut memiliki dua kondisi yang sama dan keduanya bernilai benar, sehingga teks **“Five is greater than two!”** akan tercetak dua kali. Indentasi berbeda tetap bisa dijalankan selama masih benar.


### Python Variables And Comments

In [9]:
x = 5
y = "Hello, World!"

In [10]:
#This is a comment.
print("Hello, World!")

Hello, World!


### Analisis
Kode tersebut mendefinisikan variabel `x` berisi angka 5 dan `y` berisi teks “Hello, World!”. Tanda `#` digunakan sebagai komentar. Perintah `print("Hello, World!")` menampilkan teks ke layar.


## Statements

In [11]:
print("Python is fun!")

Python is fun!


### Analisis
Kode tersebut menggunakan fungsi `print()` untuk menampilkan teks **“Python is fun!”** pada layar sebagai output.


## Many Statements

In [12]:
print("Hello World!")
print("Have a good day.")
print("Learning Python is fun!")

Hello World!
Have a good day.
Learning Python is fun!


### Analisis 
Kode tersebut menampilkan tiga baris teks secara berurutan ke layar menggunakan fungsi `print()`, sehingga setiap kalimat muncul sebagai output dalam baris terpisah.


## Semicolons (Optional, Rarely Used)

In [13]:
print("Hello"); print("How are you?"); print("Bye bye!")

Hello
How are you?
Bye bye!


### Analisis 
Kode tersebut berisi tiga perintah `print()` yang ditulis dalam satu baris menggunakan titik koma sebagai pemisah. Setiap perintah tetap dijalankan secara terpisah, sehingga tiga kalimat akan tampil berurutan di layar. Cara ini menunjukkan bahwa Python memungkinkan beberapa perintah dalam satu baris, namun biasanya tidak digunakan untuk menjaga keterbacaan kode.


## Print Text

In [14]:
print("Hello World!")

Hello World!


### Analisis
Kode `print("Hello World!")` digunakan untuk menampilkan teks **“Hello World!”** pada layar. Fungsi `print()` bertugas mencetak isi yang berada di dalam tanda kurung, dan contoh ini biasanya menjadi program awal untuk mengetes bahwa Python berjalan dengan baik serta memahami cara menampilkan output di layar.


In [16]:
print("Hello World!")
print("I am learning Python.")
print("It is awesome!")

Hello World!
I am learning Python.
It is awesome!


### Analisis
Kode tersebut berisi tiga perintah `print()` yang masing-masing menampilkan kalimat berbeda ke layar. Saat dijalankan, ketiga kalimat tersebut akan muncul secara berurutan dalam baris terpisah. Contoh ini menunjukkan cara Python menampilkan beberapa output teks sekaligus dan sering digunakan untuk latihan dasar pemrograman.


## Double Quotes

In [17]:
print("This will work!")
print('This will also work!')

This will work!
This will also work!


### Analisis 
Kode tersebut menggunakan fungsi `print()` untuk menampilkan teks ke layar. Pada contoh ini, tanda kutip ganda (`" "`) dan tanda kutip tunggal (`' '`) sama-sama digunakan untuk membuat string, dan keduanya valid dalam Python. Hasilnya, kedua baris teks akan tampil di output secara terpisah. Ini menunjukkan bahwa Python fleksibel dalam penulisan string selama pembuka dan penutup kutip sesuai.


## Print Without a New Line


In [18]:
print("Hello World!", end=" ")
print("I will print on the same line.")

Hello World! I will print on the same line.


### Analisis
Kode tersebut menggunakan `end=" "` pada perintah `print()` pertama agar tidak membuat baris baru setelah teks dicetak. Sebagai gantinya, ia menambahkan spasi di akhir. Karena itu, perintah `print()` kedua akan mencetak teks lanjutan pada baris yang sama. Hasilnya, kedua kalimat muncul dalam satu baris dengan spasi sebagai pemisah.


## Print Numbers

In [19]:
print(3)
print(358)
print(50000)

3
358
50000


### Analisis 
Kode tersebut menggunakan tiga perintah `print()` untuk menampilkan tiga nilai angka yang berbeda, yaitu 3, 358, dan 50000. Setiap angka dicetak pada baris terpisah karena setiap `print()` secara otomatis membuat baris baru. Contoh ini menunjukkan bahwa Python dapat mencetak data bertipe angka tanpa perlu tanda kutip.


In [20]:
print(3 + 3)
print(2 * 5)

6
10


### Analisis
Kode tersebut menampilkan hasil dari operasi aritmatika menggunakan fungsi `print()`. Pada baris pertama, `3 + 3` dihitung terlebih dahulu lalu hasilnya yaitu **6** ditampilkan. Pada baris kedua, `2 * 5` menghasilkan **10** dan juga ditampilkan ke layar. Ini menunjukkan bahwa Python dapat melakukan perhitungan langsung di dalam perintah `print()`.


## Mix Text and Numbers

In [21]:
print("I am", 35, "years old.")

I am 35 years old.


### Analisis 
Kode `print("I am", 35, "years old.")` menampilkan teks dan angka dalam satu perintah `print()`. Python secara otomatis menambahkan spasi di antara setiap argumen, sehingga hasilnya akan tampil sebagai kalimat lengkap di layar: **I am 35 years old.** Contoh ini menunjukkan bahwa `print()` dapat mencetak beberapa tipe data sekaligus tanpa perlu mengubah angka menjadi string terlebih dahulu.


## Creating a Comment


In [23]:
#This is a comment
print("Hello, World!")


Hello, World!


### Analisis
Kode tersebut menggunakan tanda pagar `#` untuk menulis komentar pada baris pertama. Komentar tidak akan dijalankan oleh Python dan hanya berfungsi sebagai catatan dalam kode. Baris kedua menjalankan perintah `print()` untuk menampilkan teks “Hello, World!” ke layar. Ini menunjukkan perbedaan antara komentar dan perintah yang dieksekusi dalam program Python.


In [24]:
print("Hello, World!") #This is a comment

Hello, World!


### Analisis
Kode tersebut menampilkan teks “Hello, World!” ke layar menggunakan fungsi `print()`. Setelah perintah tersebut, terdapat komentar `#This is a comment` yang tidak akan dieksekusi oleh Python. Komentar ini hanya berfungsi sebagai penjelas dalam kode, menunjukkan bahwa kita dapat menulis komentar pada akhir baris perintah tanpa mengganggu jalannya program.


In [25]:
#print("Hello, World!")
print("Cheers, Mate!")

Cheers, Mate!


### Analisis
Kode tersebut memiliki satu baris komentar dan satu baris perintah. Baris pertama diawali dengan `#` sehingga dianggap sebagai komentar dan tidak akan dijalankan oleh Python. Baris kedua menjalankan fungsi `print()` untuk menampilkan teks “Cheers, Mate!” ke layar. Hal ini menunjukkan bahwa komentar dapat digunakan untuk menonaktifkan perintah tanpa menghapusnya dari kode.


## Multiline Comments

In [26]:
#This is a comment
#written in
#more than just one line
print("Hello, World!")

Hello, World!


### Analisis 
Kode tersebut memiliki tiga baris komentar yang tidak dijalankan oleh Python, dan satu baris perintah `print()` yang menampilkan teks “Hello, World!” ke layar sebagai output.


In [27]:
"""
This is a comment
written in
more than just one line
"""
print("Hello, World!")

Hello, World!


### Analisis
Kode tersebut menggunakan tanda kutip tiga (`"""`) untuk membuat komentar atau teks yang diabaikan oleh Python dalam beberapa baris. Bagian ini tidak dieksekusi. Baris `print("Hello, World!")` akan dijalankan dan menampilkan teks “Hello, World!” ke layar.


## Creating Variables

In [28]:
x = 5
y = "John"
print(x)
print(y)

5
John


### Analisis
Kode tersebut membuat dua variabel: `x` berisi angka 5 dan `y` berisi teks “John”. Dua perintah `print()` digunakan untuk menampilkan nilai kedua variabel tersebut ke layar secara terpisah.


In [29]:
x = 4       # x is of type int
x = "Sally" # x is now of type str
print(x)

Sally


### Analisis
Kode tersebut menunjukkan bahwa variabel di Python dapat berubah tipe data. Awalnya `x` berisi angka (int), lalu diubah menjadi string “Sally”. Perintah `print(x)` akan menampilkan nilai terakhir dari `x`, yaitu Sally.


## Casting

In [30]:
x = str(3)    # x will be '3'
y = int(3)    # y will be 3
z = float(3)  # z will be 3.0

### Analisis
Kode tersebut menunjukkan konversi tipe data pada angka 3. Variabel `x` diubah menjadi string `'3'`, `y` menjadi bilangan bulat `3`, dan `z` menjadi bilangan pecahan `3.0`. Ini memperlihatkan bahwa Python dapat mengubah tipe data menggunakan fungsi `str()`, `int()`, dan `float()`.


## Get the Type

In [31]:
x = 5
y = "John"
print(type(x))
print(type(y))

<class 'int'>
<class 'str'>


### Analisis
Kode tersebut menggunakan `type()` untuk menampilkan tipe data dari variabel. Variabel `x` bertipe **int** karena berisi angka, sedangkan `y` bertipe **str** karena berisi teks “John”. Kedua tipe data tersebut kemudian dicetak ke layar.


## Single or Double Quotes

In [34]:
x = "John"
# is the same as
x = 'John'


### Analisis
Kode tersebut menunjukkan bahwa string di Python dapat ditulis menggunakan tanda kutip ganda (`" "`) maupun kutip tunggal (`' '`). Keduanya menghasilkan nilai yang sama, yaitu teks **John**.


## Case-Sensitive

In [36]:
a = 4
A = "Sally"
#A will not overwrite a

### Analisis
Kode tersebut menunjukkan bahwa Python membedakan huruf besar dan kecil pada nama variabel. Variabel `a` dan `A` dianggap berbeda, sehingga nilai `"Sally"` pada `A` tidak akan menggantikan nilai `4` pada `a`.


## Variable Names

In [37]:
myvar = "John"
my_var = "John"
_my_var = "John"
myVar = "John"
MYVAR = "John"
myvar2 = "John"

### Analisis 
Kode tersebut menunjukkan berbagai penulisan nama variabel yang **valid** dalam Python. Variabel boleh menggunakan huruf, angka (bukan di awal), garis bawah, dan kombinasi huruf besar-kecil. Semua contoh di atas merupakan nama variabel yang benar dan masing-masing dianggap berbeda oleh Python.


## Camel Case

In [38]:
myVariableName = "John"

### Analisis
Kode tersebut menggunakan gaya penamaan variabel **camelCase**, di mana setiap kata setelah kata pertama diawali huruf besar. Ini adalah salah satu metode penulisan nama variabel yang mudah dibaca dan sering digunakan dalam pemrograman.


## Pascal Case


In [39]:
MyVariableName = "John"

### Analisis 
Kode tersebut memakai gaya penamaan variabel **PascalCase**, di mana setiap kata termasuk kata pertama diawali huruf besar. Gaya ini sering digunakan untuk penamaan kelas dalam banyak bahasa pemrograman, termasuk Python.


## Snake Case

In [42]:
my_variable_name = "John"

### Analisis
Kode tersebut menggunakan gaya penamaan **snake_case**, yaitu setiap kata dipisahkan dengan garis bawah. Bentuk ini mudah dibaca dan merupakan gaya penulisan variabel yang direkomendasikan dalam Python.


## Many Values to Multiple Variables

In [43]:
x, y, z = "Orange", "Banana", "Cherry"
print(x)
print(y)
print(z)

Orange
Banana
Cherry


### Analisis
Kode tersebut melakukan **multiple assignment**, yaitu memberikan tiga nilai string ke tiga variabel sekaligus: `x` berisi “Orange”, `y` berisi “Banana”, dan `z` berisi “Cherry”. Ketiga variabel kemudian dicetak satu per satu sehingga masing-masing nilai tampil pada baris terpisah.


## One Value to Multiple Variables

In [44]:
x = y = z = "Orange"
print(x)
print(y)
print(z)

Orange
Orange
Orange


### Analisis
Kode tersebut memberikan nilai yang sama, yaitu “Orange”, ke tiga variabel sekaligus (`x`, `y`, dan `z`). Ketiga perintah `print()` kemudian menampilkan nilai tersebut, sehingga “Orange” muncul tiga kali pada output.


## Unpack a Collection

In [45]:
fruits = ["apple", "banana", "cherry"]
x, y, z = fruits
print(x)
print(y)
print(z)

apple
banana
cherry


### Analisis
Kode tersebut melakukan **unpacking list**, yaitu tiga elemen dalam list `fruits` masing-masing dimasukkan ke variabel `x`, `y`, dan `z`. Hasilnya, `x` berisi **apple**, `y` berisi **banana**, dan `z` berisi **cherry**, lalu semuanya ditampilkan ke layar dalam baris terpisah.


## Output Variables

In [46]:
x = "Python is awesome"
print(x)

Python is awesome


### Analisis
Kode tersebut menyimpan teks **“Python is awesome”** ke dalam variabel `x`, kemudian menampilkannya ke layar menggunakan perintah `print()`.


In [47]:
x = "Python"
y = "is"
z = "awesome"
print(x, y, z)

Python is awesome


### Analisis
Kode tersebut menyimpan tiga kata pada variabel `x`, `y`, dan `z`, lalu mencetak ketiganya sekaligus menggunakan `print()`. Python otomatis memberi spasi di antara setiap variabel, sehingga hasilnya tampil sebagai satu kalimat.


In [48]:
x = "Python "
y = "is "
z = "awesome"
print(x + y + z)

Python is awesome


### Analisis
Kode tersebut menggabungkan tiga string dengan operator `+`, sehingga variabel `x`, `y`, dan `z` disatukan menjadi satu teks lengkap. Hasilnya ditampilkan sebagai **“Python is awesome”**.


In [49]:
x = 5
y = 10
print(x + y)

15


### Analisis
Kode tersebut menambahkan dua nilai angka, yaitu `x = 5` dan `y = 10`. Operasi `x + y` menghasilkan **15**, kemudian hasilnya ditampilkan ke layar menggunakan `print()`.


In [50]:
x = 5
y = "John"
print(x, y)

5 John


### Analisis
Kode tersebut menyimpan angka pada `x` dan teks pada `y`. Perintah `print(x, y)` mencetak keduanya dalam satu baris, dan Python otomatis memberi spasi di antara nilai yang dicetak.


## Global Variables

In [51]:
x = "awesome"

def myfunc():
  print("Python is " + x)

myfunc()

Python is awesome


### Analisis
Kode tersebut mendefinisikan variabel `x` berisi string **“awesome”** di luar fungsi, sehingga menjadi variabel global. Fungsi `myfunc()` mencetak kalimat dengan menggabungkan teks **“Python is ”** dan nilai `x`. Saat `myfunc()` dipanggil, output yang muncul adalah **“Python is awesome”**.


In [52]:
x = "awesome"

def myfunc():
  x = "fantastic"
  print("Python is " + x)

myfunc()

print("Python is " + x)

Python is fantastic
Python is awesome


### Analisis
Kode tersebut memperlihatkan perbedaan **variabel lokal** dan **global**.
Di dalam fungsi, `x` diberi nilai **“fantastic”** sehingga saat `myfunc()` dipanggil, outputnya **“Python is fantastic”**.
Namun, variabel `x` di luar fungsi tetap **“awesome”**, sehingga baris terakhir mencetak **“Python is awesome”**.


## The global Keyword

In [53]:
def myfunc():
  global x
  x = "fantastic"

myfunc()

print("Python is " + x)

Python is fantastic


### Analisis
Program tersebut mendefinisikan fungsi `myfunc()` yang menggunakan kata kunci `global` agar variabel `x` dapat diakses di luar fungsi. Di dalam fungsi, `x` diberi nilai `"fantastic"`, dan setelah fungsi dipanggil, nilai tersebut digunakan dalam perintah `print("Python is " + x)`, sehingga menghasilkan output **"Python is fantastic"**.


In [54]:
x = "awesome"

def myfunc():
  global x
  x = "fantastic"

myfunc()

print("Python is " + x)

Python is fantastic


### Analisis
Program tersebut menetapkan `x = "awesome"`, lalu fungsi `myfunc()` mengubah nilai global `x` menjadi `"fantastic"`. Setelah fungsi dipanggil, perintah `print("Python is " + x)` menampilkan output **"Python is fantastic"**.


## Getting the Data Type

In [55]:
x = 5
print(type(x))


<class 'int'>


### Analisis
Program tersebut menetapkan nilai `x = 5`, yaitu sebuah bilangan bulat. Fungsi `type(x)` digunakan untuk mengetahui tipe data dari variabel `x`. Karena nilai `x` adalah bilangan bulat, hasil yang ditampilkan oleh program adalah **`<class 'int'>`**, yang menunjukkan bahwa `x` bertipe **integer**.


## Python Numbers

In [56]:
x = 1    # int
y = 2.8  # float
z = 1j   # complex

### Analisis
Program tersebut mendefinisikan tiga tipe data numerik: `x = 1` sebagai **integer**, `y = 2.8` sebagai **float**, dan `z = 1j` sebagai **complex** yang memiliki bagian imajiner.


In [57]:
print(type(x))
print(type(y))
print(type(z))


<class 'int'>
<class 'float'>
<class 'complex'>


### Analisis
Program tersebut menampilkan tipe data dari masing-masing variabel. Hasilnya adalah **`<class 'int'>`** untuk `x`, **`<class 'float'>`** untuk `y`, dan **`<class 'complex'>`** untuk `z`, yang menunjukkan bahwa ketiganya bertipe integer, float, dan complex.


## Int

In [58]:
x = 1
y = 35656222554887711
z = -3255522

print(type(x))
print(type(y))
print(type(z))


<class 'int'>
<class 'int'>
<class 'int'>


### Analisis
Program tersebut mendefinisikan tiga bilangan bulat, yaitu `x = 1`, `y = 35656222554887711`, dan `z = -3255522`. Semua variabel ini bertipe **integer (int)**, terlepas dari besar atau kecilnya nilai, sehingga perintah `print(type(...))` menampilkan **`<class 'int'>`** untuk ketiganya.


## Float

In [59]:
x = 1.10
y = 1.0
z = -35.59

print(type(x))
print(type(y))
print(type(z))

<class 'float'>
<class 'float'>
<class 'float'>


### Analisis
Program tersebut mendefinisikan tiga bilangan desimal: `x = 1.10`, `y = 1.0`, dan `z = -35.59`. Semua variabel bertipe **float**, sehingga perintah `print(type(...))` menampilkan **`<class 'float'>`** untuk ketiganya.


In [60]:
x = 35e3
y = 12E4
z = -87.7e100

print(type(x))
print(type(y))
print(type(z))

<class 'float'>
<class 'float'>
<class 'float'>


### Analisis 
Program tersebut mendefinisikan tiga bilangan menggunakan notasi ilmiah: `x = 35e3`, `y = 12E4`, dan `z = -87.7e100`. Semua nilai ini merupakan bilangan desimal (**float**), sehingga perintah `print(type(...))` menampilkan **`<class 'float'>`** untuk ketiganya.


## Complex

In [61]:
x = 3+5j
y = 5j
z = -5j

print(type(x))
print(type(y))
print(type(z))

<class 'complex'>
<class 'complex'>
<class 'complex'>


### Analisis
Program tersebut mendefinisikan tiga bilangan kompleks: `x = 3+5j`, `y = 5j`, dan `z = -5j`. Semua variabel bertipe **complex**, sehingga perintah `print(type(...))` menampilkan **`<class 'complex'>`** untuk ketiganya.


## Type Conversion

In [62]:
x = 1    # int
y = 2.8  # float
z = 1j   # complex

#convert from int to float:
a = float(x)

#convert from float to int:
b = int(y)

#convert from int to complex:
c = complex(x)

print(a)
print(b)
print(c)

print(type(a))
print(type(b))
print(type(c))

1.0
2
(1+0j)
<class 'float'>
<class 'int'>
<class 'complex'>


### Analisis
Program tersebut menunjukkan konversi tipe data di Python. Variabel `x` yang bertipe **int** diubah menjadi **float** pada `a = float(x)`, menghasilkan `1.0`. Variabel `y` yang bertipe **float** diubah menjadi **int** pada `b = int(y)`, menghasilkan `2` (nilai desimal dibuang). Variabel `x` juga diubah menjadi **complex** pada `c = complex(x)`, menghasilkan `(1+0j)`. Perintah `print(type(...))` menegaskan tipe baru: `a` bertipe **float**, `b` bertipe **int**, dan `c` bertipe **complex**.


## Random Number

In [63]:
import random

print(random.randrange(1, 10))

6


### Analisis
Program tersebut menggunakan modul `random` untuk menghasilkan bilangan bulat acak. Fungsi `random.randrange(1, 10)` akan mengembalikan **bilangan bulat antara 1 hingga 9** (10 tidak termasuk). Setiap kali program dijalankan, hasilnya bisa berbeda karena sifatnya acak.


## Specify a Variable Type

In [64]:
x = int(1)   # x will be 1
y = int(2.8) # y will be 2
z = int("3") # z will be 3

### Analisis
Program tersebut menunjukkan konversi ke tipe **integer** menggunakan fungsi `int()`. Variabel `x = int(1)` tetap bernilai `1`. Variabel `y = int(2.8)` dibulatkan ke bawah menjadi `2`. Variabel `z = int("3")` mengubah string `"3"` menjadi bilangan bulat `3`. Semua hasil bertipe **int**.


In [65]:
x = float(1)     # x will be 1.0
y = float(2.8)   # y will be 2.8
z = float("3")   # z will be 3.0
w = float("4.2") # w will be 4.2

### Analisis
Program tersebut menunjukkan konversi ke tipe **float** menggunakan fungsi `float()`. Variabel `x = float(1)` menjadi `1.0`. Variabel `y = float(2.8)` tetap `2.8`. Variabel `z = float("3")` mengubah string `"3"` menjadi `3.0`, dan `w = float("4.2")` mengubah string `"4.2"` menjadi `4.2`. Semua hasil bertipe **float**.


In [66]:
x = str("s1") # x will be 's1'
y = str(2)    # y will be '2'
z = str(3.0)  # z will be '3.0'

### Analisis 
Program tersebut menunjukkan konversi ke tipe **string** menggunakan fungsi `str()`. Variabel `x = str("s1")` tetap `'s1'`. Variabel `y = str(2)` mengubah bilangan `2` menjadi string `'2'`. Variabel `z = str(3.0)` mengubah bilangan desimal `3.0` menjadi string `'3.0'`. Semua hasil bertipe **str**.


## Strings

In [67]:
print("Hello")
print('Hello')

Hello
Hello


### Analisis
Program tersebut menggunakan fungsi `print()` untuk menampilkan teks `"Hello"`. Python memungkinkan penggunaan **tanda kutip ganda (" ")** maupun **tanda kutip tunggal (' ')** untuk mendefinisikan string, sehingga kedua perintah mencetak teks yang sama ke layar. Dengan demikian, output dari program ini adalah dua baris yang masing-masing berisi **Hello**.



## Quotes Inside Quotes

In [68]:
print("It's alright")
print("He is called 'Johnny'")
print('He is called "Johnny"')

It's alright
He is called 'Johnny'
He is called "Johnny"


### Analisis
Program ini menunjukkan bahwa di Python string dapat berisi tanda kutip dengan menggunakan tanda kutip tunggal atau ganda. Contohnya, `"It's alright"` menampilkan apostrof, `"He is called 'Johnny'"` menampilkan tanda kutip tunggal, dan `'He is called "Johnny"'` menampilkan tanda kutip ganda, sehingga ketiganya dicetak dengan benar di layar.


## Assign String to a Variable

In [69]:
a = "Hello"
print(a)

Hello


### Analisis
Program tersebut menyimpan string `"Hello"` ke dalam variabel `a` dan kemudian mencetaknya menggunakan fungsi `print()`. Hasil yang ditampilkan di layar adalah teks **Hello**, menunjukkan cara dasar menyimpan dan menampilkan nilai string di Python.


## Multiline Strings

In [70]:
a = """Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua."""
print(a)

Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.


### Analisis
Program tersebut menggunakan tanda kutip tiga (`""" """`) untuk membuat **multi-line string** yang mencakup beberapa baris teks. String disimpan dalam variabel `a` dan dicetak menggunakan `print()`. Hasilnya menampilkan teks persis seperti ditulis, termasuk pemisahan baris, sehingga memudahkan penulisan teks panjang dalam Python.


In [71]:
a = '''Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.'''
print(a)

Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.


### Analisis
Program tersebut menggunakan tanda kutip tiga tunggal (`''' '''`) untuk membuat **multi-line string** yang mencakup beberapa baris teks. String tersebut disimpan dalam variabel `a` dan dicetak dengan `print()`, sehingga teks ditampilkan persis seperti ditulis, termasuk pemisahan baris. Ini memudahkan penulisan teks panjang di Python.


## Strings are Arrays

In [72]:
a = "Hello, World!"
print(a[1])

e


### Analisis
Program tersebut menyimpan string `"Hello, World!"` dalam variabel `a` dan kemudian mengakses karakter pada **indeks 1** menggunakan `a[1]`. Karena Python menggunakan indeks mulai dari 0, `a[1]` merujuk pada karakter kedua, yaitu **'e'**, yang kemudian dicetak ke layar.


## Looping Through a String

In [73]:
for x in "banana":
  print(x)

b
a
n
a
n
a


### Analisis
Program tersebut menggunakan **perulangan `for`** untuk menelusuri setiap karakter dalam string `"banana"`. Setiap karakter disimpan sementara dalam variabel `x` dan dicetak dengan `print()`. Hasilnya adalah setiap huruf dari kata `"banana"` ditampilkan satu per satu pada baris terpisah.


## String Length

In [74]:
a = "Hello, World!"
print(len(a))

13


### Analisis
Program tersebut menggunakan fungsi `len()` untuk menghitung jumlah karakter dalam string `"Hello, World!"` yang disimpan di variabel `a`. Fungsi ini menghitung semua karakter termasuk spasi dan tanda baca, sehingga hasil yang dicetak adalah **13**, menunjukkan panjang string secara keseluruhan.


## Check String

In [75]:
txt = "The best things in life are free!"
print("free" in txt)

True


### Analisis
Program tersebut menggunakan operator in untuk memeriksa apakah substring "free" terdapat dalam string txt. Karena "free" memang ada di dalam teks "The best things in life are free!", hasil dari ekspresi ("free" in txt) adalah True, yang kemudian dicetak ke layar.

In [76]:
txt = "The best things in life are free!"
if "free" in txt:
  print("Yes, 'free' is present.")

Yes, 'free' is present.


### Analisis
Program tersebut menggunakan **kondisi `if`** untuk memeriksa apakah substring `"free"` terdapat dalam string `txt`. Karena `"free"` memang ada dalam teks `"The best things in life are free!"`, kondisi terpenuhi dan perintah `print()` dijalankan, sehingga output yang ditampilkan adalah **Yes, 'free' is present.**.


## Check if NOT

In [77]:
txt = "The best things in life are free!"
print("expensive" not in txt)

True


### Analisis
Program tersebut menggunakan operator **`not in`** untuk memeriksa apakah substring `"expensive"` **tidak ada** dalam string `txt`. Karena `"expensive"` memang tidak terdapat dalam teks `"The best things in life are free!"`, ekspresi `("expensive" not in txt)` bernilai **True**, yang kemudian dicetak ke layar.


In [78]:
txt = "The best things in life are free!"
if "expensive" not in txt:
  print("No, 'expensive' is NOT present.")

No, 'expensive' is NOT present.


### Analisis
Program tersebut menggunakan **kondisi `if`** untuk memeriksa apakah substring `"expensive"` **tidak ada** dalam string `txt`. Karena `"expensive"` memang tidak terdapat dalam teks `"The best things in life are free!"`, kondisi terpenuhi dan perintah `print()` dijalankan, sehingga output yang ditampilkan adalah **No, 'expensive' is NOT present.**.


## Slicing


In [79]:
b = "Hello, World!"
print(b[2:5])

llo


### Analisis
Program tersebut menggunakan **slicing** untuk mengambil sebagian string `b = "Hello, World!"` dari indeks **2 hingga 4** (indeks akhir 5 tidak termasuk). Karakter pada indeks 2, 3, dan 4 adalah `"llo"`, sehingga perintah `print(b[2:5])` menampilkan **llo** di layar.


## Slice From the Start

In [80]:
b = "Hello, World!"
print(b[:5])

Hello


### Analisis
Program tersebut menggunakan **slicing** untuk mengambil sebagian string `b = "Hello, World!"` dari awal hingga indeks **4** (indeks akhir 5 tidak termasuk). Karakter pada indeks 0 hingga 4 adalah `"Hello"`, sehingga perintah `print(b[:5])` menampilkan **Hello** di layar.


## Slice To the End

In [81]:
b = "Hello, World!"
print(b[2:])

llo, World!


### Analisis
Program tersebut menggunakan **slicing** untuk mengambil sebagian string `b = "Hello, World!"` dari indeks **2** hingga akhir string. Karakter mulai dari indeks 2 adalah `"llo, World!"`, sehingga perintah `print(b[2:])` menampilkan **llo, World!** di layar.


## Negative Indexing

In [82]:
b = "Hello, World!"
print(b[-5:-2])


orl


### Analisis
Program tersebut menggunakan **slicing dengan indeks negatif** untuk mengambil sebagian string `b = "Hello, World!"`. Indeks negatif menghitung dari akhir string, sehingga `b[-5:-2]` mengambil karakter dari posisi kelima dari belakang hingga ketiga dari belakang (indeks akhir tidak termasuk). Hasilnya adalah **"orl"**, yang kemudian dicetak ke layar.


## Upper Case

In [83]:
a = "Hello, World!"
print(a.upper())

HELLO, WORLD!


### Analisis
Program tersebut menggunakan metode `.upper()` untuk mengubah semua huruf dalam string `a = "Hello, World!"` menjadi huruf **kapital**. Hasilnya adalah **"HELLO, WORLD!"**, yang kemudian dicetak ke layar, sedangkan karakter non-huruf tetap tidak berubah.


## Lower Case

In [84]:
a = "Hello, World!"
print(a.lower())

hello, world!


### Analisis
Program tersebut menggunakan metode `.lower()` untuk mengubah semua huruf dalam string `a = "Hello, World!"` menjadi huruf **kecil**. Hasilnya adalah **"hello, world!"**, yang kemudian dicetak ke layar, sedangkan karakter non-huruf tetap tidak berubah.


## Remove Whitespace

In [85]:
a = " Hello, World! "
print(a.strip()) # returns "Hello, World!"

Hello, World!


### Analisis
Program tersebut menggunakan metode `.strip()` untuk **menghapus spasi kosong di awal dan akhir** string `a = " Hello, World! "`. Hasilnya adalah **"Hello, World!"**, yang kemudian dicetak ke layar, sementara spasi di tengah string tetap dipertahankan.


## Replace String

In [86]:
a = "Hello, World!"
print(a.replace("H", "J"))

Jello, World!


### Analisis
Program tersebut menggunakan metode `.replace()` untuk **mengganti karakter** dalam string `a = "Hello, World!"`. Metode ini mengganti semua kemunculan huruf `"H"` dengan huruf `"J"`, sehingga hasilnya adalah **"Jello, World!"**, yang kemudian dicetak ke layar.


## Split String

In [87]:
a = "Hello, World!"
print(a.split(",")) # returns ['Hello', ' World!']

['Hello', ' World!']


### Analisis
Program tersebut menggunakan metode `.split(",")` untuk **memisahkan string** `a = "Hello, World!"` menjadi daftar (list) berdasarkan **pemisah koma** `,`. Hasilnya adalah daftar **['Hello', ' World!']**, yang kemudian dicetak ke layar. Spasi setelah koma tetap dipertahankan pada elemen kedua.


## String Concatenation

In [88]:
a = "Hello"
b = "World"
c = a + b
print(c)

HelloWorld


### Analisis
Program tersebut menggunakan operator `+` untuk **menggabungkan (concatenate) dua string** `a = "Hello"` dan `b = "World"`. Hasil penggabungan disimpan dalam variabel `c`, sehingga `c` menjadi **"HelloWorld"**, yang kemudian dicetak ke layar. Perlu diperhatikan bahwa tidak ada spasi otomatis antara kedua string.


In [89]:
a = "Hello"
b = "World"
c = a + " " + b
print(c)


Hello World


### Analisis
Program tersebut menggunakan operator `+` untuk **menggabungkan (concatenate) dua string** `a = "Hello"` dan `b = "World"` dengan menambahkan spasi `" "` di antara keduanya. Hasil penggabungan disimpan dalam variabel `c`, sehingga `c` menjadi **"Hello World"**, yang kemudian dicetak ke layar.


## F-Strings

In [90]:
age = 36
txt = f"My name is John, I am {age}"
print(txt)

My name is John, I am 36


### Analisis
Program tersebut menggunakan **f-string** untuk menyisipkan nilai variabel `age` ke dalam string secara langsung. Variabel `age = 36` dimasukkan ke tempat `{age}` dalam string `txt`, sehingga hasilnya adalah **"My name is John, I am 36"**, yang kemudian dicetak ke layar.


## Placeholders and Modifiers

In [91]:
price = 59
txt = f"The price is {price} dollars"
print(txt)

The price is 59 dollars


### Analisis
Program tersebut menggunakan **f-string** untuk menyisipkan nilai variabel `price` ke dalam string. Variabel `price = 59` dimasukkan ke tempat `{price}` dalam string `txt`, sehingga hasilnya adalah **"The price is 59 dollars"**, yang kemudian dicetak ke layar.


In [92]:
price = 59
txt = f"The price is {price:.2f} dollars"
print(txt)

The price is 59.00 dollars


### Analisis
Program tersebut menggunakan **f-string** dengan format spesifier `:.2f` untuk menampilkan variabel `price` sebagai bilangan desimal dengan **2 angka di belakang koma**. Variabel `price = 59` diformat menjadi **59.00**, sehingga string `txt` menjadi **"The price is 59.00 dollars"**, yang kemudian dicetak ke layar.


In [93]:
txt = f"The price is {20 * 59} dollars"
print(txt)

The price is 1180 dollars


### Analisis
Program tersebut menggunakan **f-string** untuk menyisipkan hasil ekspresi matematika `20 * 59` langsung ke dalam string. Hasil perkalian adalah **1180**, sehingga string `txt` menjadi **"The price is 1180 dollars"**, yang kemudian dicetak ke layar.


## Escape Character

In [94]:
txt = "We are the so-called \"Vikings\" from the north."

### Analisis
Program tersebut menunjukkan penggunaan **escape character `\"`** untuk menyertakan tanda kutip ganda di dalam string. String `txt` menyimpan kalimat **We are the so-called "Vikings" from the north.**, sehingga tanda kutip di sekitar kata *Vikings* dapat muncul di output tanpa menimbulkan error.


## Boolean Values

In [95]:
print(10 > 9)
print(10 == 9)
print(10 < 9)

True
False
False


### Analisis
Program tersebut menggunakan **operator perbandingan**. `10 > 9` menghasilkan **True**, `10 == 9` menghasilkan **False**, dan `10 < 9` menghasilkan **False**, yang dicetak ke layar secara berurutan.


In [96]:
a = 200
b = 33

if b > a:
  print("b is greater than a")
else:
  print("b is not greater than a")

b is not greater than a


### Analisis
Program tersebut menggunakan **struktur percabangan `if-else`** untuk membandingkan dua variabel, `a = 200` dan `b = 33`. Kondisi `b > a` bernilai **False**, sehingga blok `else` dijalankan dan mencetak **"b is not greater than a"** ke layar.


## Evaluate Values and Variables

In [97]:
print(bool("Hello"))
print(bool(15))

True
True


### Analisis 
Program tersebut menggunakan fungsi `bool()` untuk mengubah nilai menjadi **tipe Boolean**. String `"Hello"` dianggap **True** karena bukan string kosong, dan angka `15` juga dianggap **True** karena bukan nol. Kedua hasil tersebut dicetak ke layar.


In [98]:
x = "Hello"
y = 15

print(bool(x))
print(bool(y))

True
True


### Analisis
Program tersebut menggunakan fungsi `bool()` untuk mengubah nilai menjadi **Boolean**. String `x = "Hello"` menghasilkan **True** karena bukan string kosong, dan angka `y = 15` menghasilkan **True** karena bukan nol. Kedua hasil dicetak ke layar, menunjukkan bahwa nilai non-kosong dan non-nol dianggap benar di Python.


## Most Values are True

In [99]:
bool("abc")
bool(123)
bool(["apple", "cherry", "banana"])

True

### Analisis
Program tersebut menggunakan fungsi `bool()` untuk mengubah berbagai nilai menjadi **Boolean**. String `"abc"`, angka `123`, dan list `["apple", "cherry", "banana"]` semuanya dianggap **True** karena tidak kosong atau nol. Ini menunjukkan bahwa di Python, nilai non-kosong dan non-nol dievaluasi sebagai True.


## Some Values are False

In [100]:
bool(False)
bool(None)
bool(0)
bool("")
bool(())
bool([])
bool({})

False

### Analisis
Program tersebut menggunakan fungsi `bool()` untuk mengevaluasi berbagai nilai menjadi **Boolean**. Semua nilai yang kosong atau bernilai nol, seperti `False`, `None`, `0`, string kosong `""`, tuple kosong `()`, list kosong `[]`, dan dictionary kosong `{}`, akan menghasilkan **False**. Ini menunjukkan bahwa di Python, nilai kosong atau nol dianggap salah dalam konteks logika.


In [101]:
class myclass():
  def __len__(self):
    return 0

myobj = myclass()
print(bool(myobj))


False


### Analisis
Program tersebut mendefinisikan sebuah **kelas `myclass`** dengan metode khusus `__len__()` yang selalu mengembalikan 0. Saat objek `myobj` dari kelas ini dikonversi ke Boolean menggunakan `bool()`, Python memeriksa metode `__len__()`. Karena panjang objek adalah 0, hasilnya adalah **False**, yang kemudian dicetak ke layar.


## Functions can Return a Boolean

In [102]:
def myFunction() :
  return True

print(myFunction())

True


### Analisis
Program tersebut mendefinisikan sebuah fungsi `myFunction()` yang selalu mengembalikan nilai True. Ketika fungsi dipanggil dengan `print(myFunction())`, nilai **True** yang dikembalikan oleh fungsi dicetak ke layar, menunjukkan cara dasar penggunaan `return` dalam fungsi Python.


In [103]:
def myFunction() :
  return True

if myFunction():
  print("YES!")
else:
  print("NO!")

YES!


### Analisis
Program tersebut mendefinisikan fungsi `myFunction()` yang selalu **mengembalikan True**. Fungsi ini kemudian digunakan dalam **kondisi `if`**, sehingga karena nilai yang dikembalikan True, blok `if` dijalankan dan mencetak **"YES!"** ke layar, sedangkan blok `else` diabaikan.


In [104]:
x = 200
print(isinstance(x, int))

True


### Analisis
Program tersebut menggunakan fungsi `isinstance()` untuk memeriksa apakah variabel `x = 200` adalah **tipe data `int`**. Karena `x` memang sebuah bilangan bulat, ekspresi `isinstance(x, int)` bernilai **True**, yang kemudian dicetak ke layar.


## Python Operators

In [105]:
print(10 + 5)

15


### Analisis
Program tersebut melakukan operasi **penjumlahan** antara angka `10` dan `5` menggunakan operator `+`. Hasil dari operasi ini adalah **15**, yang kemudian dicetak ke layar.


In [106]:
sum1 = 100 + 50      # 150 (100 + 50)
sum2 = sum1 + 250    # 400 (150 + 250)
sum3 = sum2 + sum2   # 800 (400 + 400)

### Analisis
Program tersebut menunjukkan penjumlahan bertahap menggunakan variabel. Pertama, `sum1 = 100 + 50` menghasilkan **150**, kemudian `sum2 = sum1 + 250` menghasilkan **400**, dan terakhir `sum3 = sum2 + sum2` menghasilkan **800**. Ini memperlihatkan cara menyimpan dan menggunakan hasil perhitungan sebelumnya dalam operasi aritmatika.


## Examples

In [107]:
x = 15
y = 4

print(x + y)
print(x - y)
print(x * y)
print(x / y)
print(x % y)
print(x ** y)
print(x // y)

19
11
60
3.75
3
50625
3


### Analisis
Program tersebut menunjukkan berbagai **operator aritmatika** di Python menggunakan `x = 15` dan `y = 4`. Operator `+` menghasilkan penjumlahan (**19**), `-` pengurangan (**11**), `*` perkalian (**60**), `/` pembagian (**3.75**), `%` sisa bagi (**3**), `**` perpangkatan (**50625**), dan `//` pembagian lantai (**3**). Hasil dari setiap operasi dicetak ke layar.


## Division in Python

In [108]:
x = 12
y = 5

print(x / y)

2.4


### Analisis
Program tersebut menggunakan operator `/` untuk melakukan pembagian antara `x = 12` dan `y = 5`. Hasil pembagian adalah **2.4**, yang dicetak ke layar sebagai bilangan desimal (float).


In [109]:
x = 12
y = 5

print(x // y)

2


### Analisis
Program tersebut menggunakan operator `//` untuk melakukan **pembagian lantai (floor division)** antara `x = 12` dan `y = 5`. Hasilnya adalah **2**, karena pembagian lantai mengabaikan sisa dan hanya mengambil bagian bulat dari hasil pembagian, yang kemudian dicetak ke layar.


## The Walrus Operator

In [110]:
numbers = [1, 2, 3, 4, 5]
count = len(numbers)
if count > 3:
    print(f"List has {count} elements")

if (count := len(numbers)) > 3:
    print(f"List has {count} elements")

List has 5 elements
List has 5 elements


### Analisis
Program tersebut menunjukkan dua cara untuk menghitung dan memeriksa jumlah elemen dalam list `numbers = [1, 2, 3, 4, 5]`. Pada blok pertama, `len(numbers)` disimpan di variabel `count` lalu dibandingkan dengan 3; karena jumlah elemen **5 > 3**, dicetak **"List has 5 elements"**. Pada blok kedua, operator **walrus `:=`** digunakan untuk menghitung panjang list langsung dalam kondisi `if`; hasilnya sama, dicetak **"List has 5 elements"**, tetapi lebih ringkas.


## Examples

In [111]:
x = 5
y = 3

print(x == y)
print(x != y)
print(x > y)
print(x < y)
print(x >= y)
print(x <= y)

False
True
True
False
True
False


### Analisis
Program tersebut menggunakan berbagai **operator perbandingan** untuk membandingkan `x = 5` dan `y = 3`. `x == y` mengecek kesamaan (**False**), `x != y` mengecek ketidaksetaraan (**True**), `x > y` mengecek apakah x lebih besar (**True**), `x < y` mengecek apakah x lebih kecil (**False**), `x >= y` mengecek lebih besar atau sama dengan (**True**), dan `x <= y` mengecek lebih kecil atau sama dengan (**False**). Semua hasil dicetak ke layar secara berurutan.


## Chaining Comparison Operators

In [112]:
x = 5

print(1 < x < 10)

print(1 < x and x < 10)

True
True


### Analisis
Program tersebut menunjukkan dua cara untuk memeriksa apakah `x = 5` berada di antara 1 dan 10. Pernyataan rantai perbandingan `1 < x < 10` langsung menghasilkan **True**, sedangkan ekspresi logika `1 < x and x < 10` menggunakan operator **and** untuk mencapai hasil yang sama, yaitu **True**. Kedua metode dicetak ke layar dan menghasilkan hasil identik.


## Logical Operators

In [113]:
x = 5

print(x > 0 and x < 10)

True


### Analisis
Program tersebut menggunakan operator logika **and** untuk memeriksa dua kondisi sekaligus: apakah `x = 5` lebih besar dari 0 dan lebih kecil dari 10. Karena kedua kondisi benar, ekspresi menghasilkan **True**, yang kemudian dicetak ke layar.


In [114]:
x = 5

print(x < 5 or x > 10)

False


### Analisis
Program tersebut menggunakan operator logika **or** untuk memeriksa dua kondisi: apakah `x = 5` lebih kecil dari 5 atau lebih besar dari 10. Karena kedua kondisi salah, ekspresi menghasilkan **False**, yang kemudian dicetak ke layar.


In [115]:
x = 5

print(not(x > 3 and x < 10))

False


### Analisis
Program tersebut menggunakan operator logika **not** untuk membalik hasil dari ekspresi `x > 3 and x < 10`. Karena `x = 5` memang lebih besar dari 3 dan lebih kecil dari 10, ekspresi di dalam `not` bernilai **True**, sehingga setelah dibalik menjadi **False**, yang kemudian dicetak ke layar.


## Identity Operators

In [116]:
x = ["apple", "banana"]
y = ["apple", "banana"]
z = x

print(x is z)
print(x is y)
print(x == y)

True
False
True


### Analisis
Program tersebut membandingkan list menggunakan **operator `is`** dan **operator `==`**. `x is z` bernilai **True** karena `z` adalah referensi ke objek yang sama dengan `x`. `x is y` bernilai **False** karena `y` adalah list terpisah meskipun memiliki isi sama. `x == y` bernilai **True** karena operator `==` membandingkan **isi** kedua list, bukan identitas objek. Hasilnya dicetak ke layar.


In [117]:
x = ["apple", "banana"]
y = ["apple", "banana"]

print(x is not y)

True


### Analisis
Program tersebut menggunakan operator `is not` untuk memeriksa apakah dua list `x` dan `y` **bukan objek yang sama** di memori. Meskipun kedua list memiliki isi yang sama, mereka adalah objek berbeda, sehingga ekspresi `x is not y` menghasilkan **True**, yang kemudian dicetak ke layar.


## Difference Between is and ==

In [118]:
x = [1, 2, 3]
y = [1, 2, 3]

print(x == y)
print(x is y)

True
False


### Analisis
Program tersebut membandingkan dua list `x` dan `y` menggunakan **operator `==`** dan **`is`**. `x == y` bernilai **True** karena membandingkan **isi** kedua list yang sama, sedangkan `x is y` bernilai **False** karena kedua list adalah objek **berbeda** di memori. Hasil kedua perbandingan dicetak ke layar.


## Membership Operators

In [119]:
fruits = ["apple", "banana", "cherry"]

print("banana" in fruits)

True


### Analisis
Program tersebut menggunakan operator **`in`** untuk memeriksa apakah string `"banana"` ada dalam list `fruits = ["apple", "banana", "cherry"]`. Karena `"banana"` memang ada di dalam list, ekspresi menghasilkan **True**, yang kemudian dicetak ke layar.


In [120]:
fruits = ["apple", "banana", "cherry"]

print("pineapple" not in fruits)

True


### Analisis
Program tersebut menggunakan operator **`not in`** untuk memeriksa apakah string `"pineapple"` **tidak ada** dalam list `fruits = ["apple", "banana", "cherry"]`. Karena `"pineapple"` memang tidak ada di dalam list, ekspresi menghasilkan **True**, yang kemudian dicetak ke layar.


## Membership in Strings

In [121]:
text = "Hello World"

print("H" in text)
print("hello" in text)
print("z" not in text)

True
False
True


### Analisis 
Program tersebut memeriksa keberadaan karakter atau substring dalam string `text = "Hello World"` menggunakan operator **`in`** dan **`not in`**. `"H" in text` bernilai **True** karena huruf H ada di teks, `"hello" in text` bernilai **False** karena Python **membedakan huruf besar dan kecil**, dan `"z" not in text` bernilai **True** karena huruf z tidak ada di teks. Semua hasil dicetak ke layar.


## Bitwise Operators

In [122]:
print(6 & 3)

2


### Analisis
Program tersebut menggunakan operator **bitwise AND (`&`)** untuk membandingkan bilangan biner `6` (110₂) dan `3` (011₂). Operator `&` mengembalikan 1 hanya jika kedua bit bernilai 1, sehingga hasilnya adalah **2** (010₂), yang kemudian dicetak ke layar.


In [123]:
print(6 | 3)

7


### Analisis
Program tersebut menggunakan operator **bitwise OR (`|`)** untuk membandingkan bilangan biner `6` (110₂) dan `3` (011₂). Operator `|` mengembalikan 1 jika salah satu atau kedua bit bernilai 1, sehingga hasilnya adalah **7** (111₂), yang kemudian dicetak ke layar.


In [124]:
print(6 ^ 3)


5


### Analisis
berikan analisis yang jelas dan singkat dalam bentuk paragraf

## Operator Precedence

In [125]:
print((6 + 3) - (6 + 3))

0


### Analisis
Program tersebut melakukan operasi aritmatika sederhana: terlebih dahulu menjumlahkan `6 + 3` menjadi **9**, kemudian mengurangkan hasil yang sama `9 - 9`, sehingga hasil akhirnya adalah **0**, yang kemudian dicetak ke layar.


In [126]:
print(100 + 5 * 3)

115


### Analisis
Program tersebut melakukan operasi aritmatika dengan memperhatikan **prioritas operator**. Perkalian `5 * 3` dihitung terlebih dahulu menjadi **15**, kemudian dijumlahkan dengan `100`, menghasilkan **115**, yang kemudian dicetak ke layar.


## Left-to-Right Evaluation

In [127]:
print(5 + 4 - 7 + 3)

5


### Analisis 
Program tersebut melakukan operasi aritmatika **penjumlahan dan pengurangan berurutan** dari kiri ke kanan. Perhitungannya: `5 + 4 = 9`, `9 - 7 = 2`, `2 + 3 = 5`, sehingga hasil akhirnya adalah **5**, yang kemudian dicetak ke layar.


## Python Conditions and If statements

In [1]:
a = 33
b = 200
if b > a:
  print("b is greater than a")


b is greater than a


### Analisis
Program tersebut membandingkan dua nilai, yaitu `a = 33` dan `b = 200`. Karena `b` lebih besar dari `a`, kondisi `if b > a` bernilai benar sehingga program menampilkan pesan **“b is greater than a”**.


## How If Statements Work

In [2]:
number = 15
if number > 0:
  print("The number is positive")

The number is positive


### Analisis
Program tersebut mendefinisikan variabel `number` dengan nilai 15. Kemudian, pernyataan `if number > 0:` memeriksa apakah nilai tersebut lebih besar dari nol. Karena 15 merupakan bilangan positif, kondisi tersebut bernilai benar, sehingga program menampilkan pesan **“The number is positive”** pada layar.


## Multiple Statements in If Block

In [3]:
age = 20
if age >= 18:
  print("You are an adult")
  print("You can vote")
  print("You have full legal rights")

You are an adult
You can vote
You have full legal rights


### Analisis
Program tersebut memeriksa apakah `age` bernilai 18 atau lebih. Karena `age = 20`, kondisi terpenuhi, sehingga program menampilkan pesan **“You are an adult”**, **“You can vote”**, dan **“You have full legal rights”**.


## Using Variables in Conditions

In [4]:
is_logged_in = True
if is_logged_in:
  print("Welcome back!")

Welcome back!


### Analisis
Program tersebut memeriksa apakah variabel `is_logged_in` bernilai True. Karena nilainya memang True, kondisi `if is_logged_in:` terpenuhi, sehingga program menampilkan pesan **“Welcome back!”** di layar.


## The Elif Keyword

In [5]:
a = 33
b = 33
if b > a:
  print("b is greater than a")
elif a == b:
  print("a and b are equal")

a and b are equal


### Analisis
Program tersebut membandingkan dua variabel, `a` dan `b`, yang keduanya bernilai 33. Kondisi pertama `if b > a` tidak terpenuhi karena nilainya sama. Program kemudian memeriksa kondisi `elif a == b`, yang bernilai benar, sehingga program menampilkan pesan **“a and b are equal”**.


## Multiple Elif Statements

In [6]:
score = 75

if score >= 90:
  print("Grade: A")
elif score >= 80:
  print("Grade: B")
elif score >= 70:
  print("Grade: C")
elif score >= 60:
  print("Grade: D")

Grade: C


### Analisis
Program tersebut memeriksa nilai `score` yang bernilai 75 untuk menentukan grade. Kondisi pertama dan kedua tidak terpenuhi karena 75 kurang dari 90 dan 80. Namun, kondisi `elif score >= 70` bernilai benar, sehingga program menampilkan pesan **“Grade: C”**.


## How Elif Works

In [7]:
age = 25

if age < 13:
  print("You are a child")
elif age < 20:
  print("You are a teenager")
elif age < 65:
  print("You are an adult")
elif age >= 65:
  print("You are a senior")

You are an adult


### Analisis
Program ini memeriksa nilai `age` yang bernilai 25 untuk menentukan kategori usia. Karena 25 tidak kurang dari 13 atau 20, kondisi pertama dan kedua tidak terpenuhi. Kondisi `elif age < 65` bernilai benar, sehingga program menampilkan pesan **“You are an adult”**.


## When to Use Elif

In [8]:
day = 3

if day == 1:
  print("Monday")
elif day == 2:
  print("Tuesday")
elif day == 3:
  print("Wednesday")
elif day == 4:
  print("Thursday")
elif day == 5:
  print("Friday")
elif day == 6:
  print("Saturday")
elif day == 7:
  print("Sunday")

Wednesday


### Analisis
Program ini memeriksa nilai variabel `day` yang bernilai 3 untuk menentukan hari dalam seminggu. Kondisi `if` dan `elif` sebelumnya tidak terpenuhi, tetapi `elif day == 3` bernilai benar, sehingga program menampilkan pesan **“Wednesday”**.


## The Else Keyword

In [9]:
a = 200
b = 33
if b > a:
  print("b is greater than a")
elif a == b:
  print("a and b are equal")
else:
  print("a is greater than b")

a is greater than b


### Analisis
Program ini membandingkan dua variabel, `a = 200` dan `b = 33`. Kondisi pertama `if b > a` tidak terpenuhi karena 33 lebih kecil dari 200, dan kondisi `elif a == b` juga tidak terpenuhi karena nilainya berbeda. Oleh karena itu, blok `else` dijalankan, sehingga program menampilkan pesan **“a is greater than b”**.


## Else Without Elif

In [10]:
a = 200
b = 33
if b > a:
  print("b is greater than a")
else:
  print("b is not greater than a")

b is not greater than a


### Analisis
Program ini membandingkan dua variabel, `a = 200` dan `b = 33`. Karena kondisi `if b > a` tidak terpenuhi—karena 33 lebih kecil dari 200—blok `else` dijalankan, sehingga program menampilkan pesan **“b is not greater than a”**.


## How Else Works

In [11]:
number = 7

if number % 2 == 0:
  print("The number is even")
else:
  print("The number is odd")

The number is odd


### Analisis
Program ini memeriksa apakah variabel `number` yang bernilai 7 adalah bilangan genap atau ganjil. Kondisi `if number % 2 == 0` memeriksa sisa pembagian 7 dengan 2. Karena 7 % 2 menghasilkan 1, kondisi tersebut salah, sehingga blok `else` dijalankan dan program menampilkan pesan **“The number is odd”**.


## Complete If-Elif-Else Chain

In [12]:
temperature = 22

if temperature > 30:
  print("It's hot outside!")
elif temperature > 20:
  print("It's warm outside")
elif temperature > 10:
  print("It's cool outside")
else:
  print("It's cold outside!")

It's warm outside


### Analisis
Program ini memeriksa nilai variabel `temperature` yang bernilai 22 untuk menentukan kondisi cuaca. Kondisi pertama `if temperature > 30` tidak terpenuhi, tetapi kondisi `elif temperature > 20` bernilai benar karena 22 lebih besar dari 20. Oleh karena itu, program menampilkan pesan **“It's warm outside”**.


## Else as Fallback

In [13]:
username = "Emil"

if len(username) > 0:
  print(f"Welcome, {username}!")
else:
  print("Error: Username cannot be empty")


Welcome, Emil!


### Analisis
Program ini memeriksa panjang variabel `username` yang bernilai "Emil". Karena panjangnya lebih dari 0, kondisi `if len(username) > 0` bernilai benar, sehingga program menampilkan pesan **“Welcome, Emil!”**. Jika `username` kosong, program akan menampilkan pesan error.


## Short Hand If

In [14]:
a = 5
b = 2
if a > b: print("a is greater than b")

a is greater than b


### Analisis
Program ini membandingkan dua variabel, `a = 5` dan `b = 2`. Karena 5 lebih besar dari 2, kondisi `if a > b` bernilai benar, sehingga program menampilkan pesan **“a is greater than b”**.


## Short Hand If ... Else

In [15]:
a = 2
b = 330
print("A") if a > b else print("B")

B


### Analisis
Program ini menggunakan **conditional expression** (ternary operator) untuk membandingkan `a = 2` dan `b = 330`. Karena 2 tidak lebih besar dari 330, kondisi `a > b` bernilai salah, sehingga program mengeksekusi bagian `else` dan menampilkan pesan **“B”**.


## Assign a Value With If ... Else

In [16]:
a = 10
b = 20
bigger = a if a > b else b
print("Bigger is", bigger)

Bigger is 20


### Analisis
Program ini menggunakan **conditional expression** untuk menentukan nilai yang lebih besar antara `a = 10` dan `b = 20`. Karena 10 tidak lebih besar dari 20, nilai `bigger` diisi dengan `b`, sehingga program menampilkan pesan **“Bigger is 20”**.


## Multiple Conditions on One Line

In [17]:
a = 330
b = 330
print("A") if a > b else print("=") if a == b else print("B")

=


### Analisis
Program ini menggunakan **nested conditional expression** untuk membandingkan `a = 330` dan `b = 330`. Karena `a > b` salah dan `a == b` benar, program mengeksekusi kondisi tengah dan menampilkan pesan **“=”**, menandakan kedua nilai sama.


### Practical Examples

In [18]:
x = 15
y = 20
max_value = x if x > y else y
print("Maximum value:", max_value)

Maximum value: 20


### Analisis
Program ini menggunakan **conditional expression** untuk menentukan nilai maksimum antara `x = 15` dan `y = 20`. Karena 15 tidak lebih besar dari 20, variabel `max_value` diisi dengan `y`, sehingga program menampilkan pesan **“Maximum value: 20”**.


In [20]:
username = ""
display_name = username if username else "Guest"
print("Welcome,", display_name)

Welcome, Guest


### Analisis
Program ini menggunakan **conditional expression** untuk menentukan nama yang akan ditampilkan. Karena `username` kosong, nilai `display_name` diisi dengan **“Guest”**, sehingga program menampilkan pesan **“Welcome, Guest”**.


## The and Operator

In [21]:
a = 200
b = 33
c = 500
if a > b and c > a:
  print("Both conditions are True")

Both conditions are True


### Analisis
Program ini memeriksa dua kondisi sekaligus menggunakan operator logika **and**: apakah `a > b` dan `c > a`. Karena `a = 200` lebih besar dari `b = 33` dan `c = 500` lebih besar dari `a = 200`, kedua kondisi bernilai benar. Akibatnya, program mengeksekusi blok `if` dan menampilkan pesan **“Both conditions are True”**.


## The or Operator

In [22]:
a = 200
b = 33
c = 500
if a > b or a > c:
  print("At least one of the conditions is True")

At least one of the conditions is True


### Analisis
Program ini memeriksa dua kondisi menggunakan operator logika **or**: apakah `a > b` atau `a > c`. Karena `a = 200` lebih besar dari `b = 33`, salah satu kondisi bernilai benar, sehingga blok `if` dijalankan dan program menampilkan pesan **“At least one of the conditions is True”**.


## The not Operator

In [23]:
a = 33
b = 200
if not a > b:
  print("a is NOT greater than b")

a is NOT greater than b


### Analisis
Program ini menggunakan operator logika **not** untuk membalik kondisi `a > b`. Karena `a = 33` tidak lebih besar dari `b = 200`, kondisi `not a > b` bernilai benar, sehingga program mengeksekusi blok `if` dan menampilkan pesan **“a is NOT greater than b”**.


## Combining Multiple Operators

In [24]:
age = 25
is_student = False
has_discount_code = True

if (age < 18 or age > 65) and not is_student or has_discount_code:
  print("Discount applies!")

Discount applies!


### Analisis
Program ini memeriksa beberapa kondisi untuk menentukan apakah diskon berlaku. Variabel `age = 25`, `is_student = False`, dan `has_discount_code = True`. Karena operator **or** dan **and** digunakan, kondisi `(age < 18 or age > 65) and not is_student` bernilai False, tetapi karena `has_discount_code = True`, keseluruhan ekspresi menjadi True. Akibatnya, program mengeksekusi blok `if` dan menampilkan pesan **“Discount applies!”**.


## Using Parentheses for Clarity

In [25]:
temperature = 25
is_raining = False
is_weekend = True

if (temperature > 20 and not is_raining) or is_weekend:
  print("Great day for outdoor activities!")

Great day for outdoor activities!


### Analisis
Program ini memeriksa kondisi untuk menentukan apakah hari tersebut cocok untuk aktivitas luar ruangan. Variabel `temperature = 25`, `is_raining = False`, dan `is_weekend = True`. Karena `temperature > 20` dan `not is_raining` bernilai True, dan juga `is_weekend = True`, keseluruhan ekspresi logika menjadi True. Akibatnya, program mengeksekusi blok `if` dan menampilkan pesan **“Great day for outdoor activities!”**.


## More Examples

In [28]:
username = "Tobias"
password = "secret123"
is_verified = True

if username and password and is_verified:
  print("Login successful")
else:
  print("Login failed")

Login successful


### Analisis
Program ini memeriksa tiga kondisi sekaligus untuk menentukan keberhasilan login: apakah `username` dan `password` memiliki nilai, serta apakah `is_verified` bernilai True. Karena semua kondisi terpenuhi (`username = "Tobias"`, `password = "secret123"`, dan `is_verified = True`), blok `if` dijalankan, sehingga program menampilkan pesan **“Login successful”**.


In [29]:
score = 85

if score >= 0 and score <= 100:
  print("Valid score")
else:
  print("Invalid score")

Valid score


### Analisis
Program ini memeriksa apakah nilai `score = 85` berada dalam rentang yang valid, yaitu 0 hingga 100. Karena 85 termasuk dalam rentang tersebut, kondisi `if score >= 0 and score <= 100` bernilai benar, sehingga program mengeksekusi blok `if` dan menampilkan pesan **“Valid score”**.


## Nested If Statements

In [30]:
x = 41

if x > 10:
  print("Above ten,")
  if x > 20:
    print("and also above 20!")
  else:
    print("but not above 20.")


Above ten,
and also above 20!


### Analisis
Program ini menggunakan **nested if** untuk memeriksa nilai `x = 41`. Kondisi pertama `if x > 10` bernilai benar, sehingga program mengeksekusi blok pertama dan menampilkan **“Above ten,”**. Kemudian, kondisi di dalam blok tersebut `if x > 20` juga bernilai benar, sehingga program menampilkan tambahan pesan **“and also above 20!””**.


## How Nested If Works

In [31]:
age = 25
has_license = True

if age >= 18:
  if has_license:
    print("You can drive")
  else:
    print("You need a license")
else:
  print("You are too young to drive")

You can drive


### Analisis
Program ini menggunakan **nested if** untuk menentukan apakah seseorang bisa mengemudi. Variabel `age = 25` dan `has_license = True`. Karena usia 25 lebih besar dari atau sama dengan 18, kondisi pertama terpenuhi. Selanjutnya, karena `has_license` bernilai True, program mengeksekusi blok dalam dan menampilkan pesan **“You can drive”**.


## Multiple Levels of Nesting

In [32]:
score = 85
attendance = 90
submitted = True

if score >= 60:
  if attendance >= 80:
    if submitted:
      print("Pass with good standing")
    else:
      print("Pass but missing assignment")
  else:
    print("Pass but low attendance")
else:
  print("Fail")

Pass with good standing


### Analisis 
Program ini menggunakan **nested if** untuk mengevaluasi kelulusan berdasarkan tiga kriteria: `score = 85`, `attendance = 90`, dan `submitted = True`. Karena nilai `score` ≥ 60, hadir ≥ 80, dan tugas sudah dikumpulkan (`submitted = True`), semua kondisi terpenuhi. Akibatnya, program mengeksekusi blok terdalam dan menampilkan pesan **“Pass with good standing”**.


## Nested If vs Logical Operators

In [33]:
temperature = 25
is_sunny = True

if temperature > 20:
  if is_sunny:
    print("Perfect beach weather!")

Perfect beach weather!


### Analisis
Program ini menggunakan **nested if** untuk memeriksa kondisi cuaca. Variabel `temperature = 25` dan `is_sunny = True`. Karena suhu lebih dari 20 dan sedang cerah, kedua kondisi terpenuhi, sehingga program mengeksekusi blok terdalam dan menampilkan pesan **“Perfect beach weather!”**.


In [34]:
temperature = 25
is_sunny = True

if temperature > 20 and is_sunny:
  print("Perfect beach weather!")

Perfect beach weather!


### Analisis
Program ini memeriksa dua kondisi sekaligus menggunakan operator logika **and**: apakah `temperature > 20` dan `is_sunny` bernilai True. Karena `temperature = 25` dan `is_sunny = True`, kedua kondisi terpenuhi, sehingga program mengeksekusi blok `if` dan menampilkan pesan **“Perfect beach weather!”**.


## More Examples

In [35]:
username = "Emil"
password = "python123"
is_active = True

if username:
  if password:
    if is_active:
      print("Login successful")
    else:
      print("Account is not active")
  else:
    print("Password required")
else:
  print("Username required")

Login successful


### Analisis
Program ini menggunakan **nested if** untuk memeriksa kondisi login berdasarkan tiga variabel: `username = "Emil"`, `password = "python123"`, dan `is_active = True`. Karena semua variabel memiliki nilai yang valid dan akun aktif, semua kondisi terpenuhi, sehingga program mengeksekusi blok terdalam dan menampilkan pesan **“Login successful”**.


In [36]:
score = 92
extra_credit = 5

if score >= 90:
  if extra_credit > 0:
    print("A+ grade")
  else:
    print("A grade")
elif score >= 80:
  print("B grade")
else:
  print("C grade or below")

A+ grade


### Analisis
Program ini menggunakan **nested if** untuk menentukan nilai berdasarkan `score = 92` dan `extra_credit = 5`. Karena skor ≥ 90 dan nilai tambahan lebih dari 0, kedua kondisi terpenuhi, sehingga program mengeksekusi blok terdalam dan menampilkan pesan **“A+ grade”**.


## The pass Statement

In [39]:
a = 33
b = 200

if b > a:
  pass

### Analisis
Program ini memeriksa apakah `b = 200` lebih besar dari `a = 33`. Karena kondisi `if b > a` bernilai benar, Python mengeksekusi pernyataan **`pass`**, yang berarti tidak melakukan apa-apa. Program ini hanya memeriksa kondisi tanpa menghasilkan output atau aksi apapun.


## pass in Development

In [40]:
age = 16

if age < 18:
  pass # TODO: Add underage logic later
else:
  print("Access granted")

### Analisis
Program ini memeriksa nilai `age = 16` untuk menentukan akses. Karena usia kurang dari 18, kondisi `if age < 18` bernilai benar, tetapi bloknya hanya berisi **`pass`**, sehingga tidak ada aksi yang dilakukan saat ini. Jika usia 18 atau lebih, program akan mengeksekusi blok `else` dan menampilkan pesan **“Access granted”**.


## pass vs Comments

In [41]:
score = 85

if score > 90:
  pass # This is excellent
print("Score processed")

Score processed


### Analisis
Program ini memeriksa apakah `score = 85` lebih besar dari 90. Karena kondisi tersebut salah, blok `if` berisi **`pass`** tidak dijalankan, sehingga tidak ada aksi dari `if`. Selanjutnya, program mengeksekusi perintah berikutnya dan menampilkan pesan **“Score processed”**, menandakan bahwa skor telah diproses.


## pass with Multiple Conditions

In [44]:
value = 50

if value < 0:
  print("Negative value")
elif value == 0:
  pass # Zero case - no action needed
else:
  print("Positive value")

Positive value


### Analisis
Program ini memeriksa nilai `value = 50` untuk menentukan sifatnya. Karena 50 tidak kurang dari 0 dan tidak sama dengan 0, kondisi `else` terpenuhi, sehingga program mengeksekusi blok tersebut dan menampilkan pesan **“Positive value”**. Blok `pass` pada kondisi nol tidak dijalankan karena nilai bukan nol.


## pass in Other Contexts

In [45]:
def calculate_discount(price):
  pass # TODO: Implement discount logic

# Function exists but doesn't do anything yet

### Analisis
Fungsi `calculate_discount(price)` didefinisikan tetapi saat ini tidak melakukan apa-apa karena hanya berisi **`pass`**. Ini berarti fungsi tersebut ada dan dapat dipanggil, namun belum memiliki logika untuk menghitung diskon. Baris komentar menunjukkan bahwa logika diskon akan ditambahkan di masa depan.


## The Python Match Statement

In [47]:
day = 4
match day:
  case 1:
    print("Monday")
  case 2:
    print("Tuesday")
  case 3:
    print("Wednesday")
  case 4:
    print("Thursday")
  case 5:
    print("Friday")
  case 6:
    print("Saturday")
  case 7:
    print("Sunday")

Thursday


### Analisis 
Program ini menggunakan **struktur match-case** untuk menentukan hari dalam seminggu berdasarkan nilai `day = 4`. Karena nilai `day` sama dengan 4, program mengeksekusi blok `case 4` dan menampilkan pesan **“Thursday”**. Struktur ini mirip dengan penggunaan `if-elif-else` tetapi lebih rapi untuk banyak kondisi diskrit.


## Default Value

In [48]:
day = 4
match day:
  case 6:
    print("Today is Saturday")
  case 7:
    print("Today is Sunday")
  case _:
    print("Looking forward to the Weekend")

Looking forward to the Weekend


### Analisis
Program ini menggunakan **struktur match-case** untuk mengevaluasi nilai `day = 4`. Karena 4 tidak sesuai dengan `case 6` atau `case 7`, program mengeksekusi `case _`, yang berfungsi sebagai **default**, sehingga menampilkan pesan **“Looking forward to the Weekend”**.


## Combine Values

In [49]:
day = 4
match day:
  case 1 | 2 | 3 | 4 | 5:
    print("Today is a weekday")
  case 6 | 7:
    print("I love weekends!")

Today is a weekday


### Analisis
Program ini menggunakan **struktur match-case** dengan operator **or (`|`)** untuk mengelompokkan beberapa nilai. Karena `day = 4` termasuk dalam `case 1 | 2 | 3 | 4 | 5`, program mengeksekusi blok ini dan menampilkan pesan **“Today is a weekday”**, menandakan bahwa hari tersebut adalah hari kerja.


## If Statements as Guards

In [50]:
month = 5
day = 4
match day:
  case 1 | 2 | 3 | 4 | 5 if month == 4:
    print("A weekday in April")
  case 1 | 2 | 3 | 4 | 5 if month == 5:
    print("A weekday in May")
  case _:
    print("No match")

A weekday in May


### Analisis
Program ini menggunakan **match-case** dengan **guard condition** (`if`) untuk memeriksa kombinasi hari dan bulan. Variabel `day = 4` dan `month = 5`, sehingga cocok dengan `case 1 | 2 | 3 | 4 | 5 if month == 5`. Akibatnya, program mengeksekusi blok ini dan menampilkan pesan **“A weekday in May”**.